src/anomalib/trainer/utils/metrics_manager.py

```diff
@@ -0,0 +1,154 @@
+
+        Args:
+            task (TaskType): Task type
+        """
+        if self.anomaly_module is None:
+            raise MisconfigurationException("Anomaly module is not available yet.")
+
+        if not hasattr(self.anomaly_module, "image_metrics"):
+            image_metric_names = [] if self.image_metric_names is None else self.image_metric_names
+
+            pixel_metric_names: list[str]
+            if self.pixel_metric_names is None:
+                pixel_metric_names = []
+            elif self.trainer.task_type == TaskType.CLASSIFICATION:
+                pixel_metric_names = []
+                logger.warning(
+                    "Cannot perform pixel-level evaluation when task type is classification. "
+                    "Ignoring the following pixel-level metrics: %s",
+                    self.pixel_metric_names,
+                )
+            else:
+                pixel_metric_names = self.pixel_metric_names
+
+            self.image_metrics = create_metric_collection(image_metric_names, "image_")
+            self.pixel_metrics = create_metric_collection(pixel_metric_names, "pixel_")
+
+            self.image_metrics.set_threshold(self.anomaly_module.image_threshold.value)
+            self.pixel_metrics.set_threshold(self.anomaly_module.pixel_threshold.value)
+
+    def set_threshold(self) -> None:
+        """Sets threshold."""
+
+        if self.trainer.state.stage in (TrainerFn.TESTING, TrainerFn.PREDICTING):
+            image_metrics_threshold = 0.5
+            pixel_metrics_threshold = 0.5
+        else:
+            image_metrics_threshold = self.anomaly_module.image_threshold.value
+            pixel_metrics_threshold = self.anomaly_module.pixel_threshold.value
+
+        if self.image_metrics is not None:
+            self.image_metrics.set_threshold(image_metrics_threshold)
+        if self.pixel_metrics is not None:
+            self.pixel_metrics.set_threshold(pixel_metrics_threshold)
+
+    def _update_metrics(self, outputs: STEP_OUTPUT | EPOCH_OUTPUT | List[EPOCH_OUTPUT]):
+        """Update metrics.
+
+        Args:
+            outputs (EPOCH_OUTPUT | List[EPOCH_OUTPUT]): Step outputs
+        """
+        if isinstance(outputs, list):
+            for output in outputs:
+                self._update_metrics(output)
+        else:
+            self.image_metrics.cpu()
+            self.image_metrics.update(outputs["pred_scores"], outputs["label"].int())
+            if "mask" in outputs.keys() and "anomaly_maps" in outputs.keys():
+                self.pixel_metrics.cpu()
+                self.pixel_metrics.update(outputs["anomaly_maps"], outputs["mask"].int())
+
+    def compute(self, outputs: EPOCH_OUTPUT | List[EPOCH_OUTPUT]) -> None:
+        """Computes the metrics.
+
+        1. Update metrics
+        2. Compute metrics
+
+        Args:
+            outputs (EPOCH_OUTPUT): Epoch outputs
+        """
+        self._update_metrics(outputs)
```

Comment:
The threshold is computed at the end of the epoch. The metrics then use this threshold when compute is called. When I call `update` at the end of the batch and call `compute` at the end of the epoch, I get poor F1 scores.

Reply:
Hmm, I see. It's because the F1Score updates the tp, fp, tn, fn counts during the update call, and for this it uses the threshold. So it seems there is no way around updating the metrics at the end of the epoch, at least for the validation sequence. For the test sequence we could still do it at the end of each batch. I feel it would be a bit better in line with the intended use of the torchmetrics classes. The drawback would be that we'll have slightly different implementation betwen val and test.